File: test/test_gravity.py
import unittest

import numpy as np
import sys

from gdf_reader import *
from gfc_reader import *

from gravity import Gravity

def potential(cs, r0, mu, r):
    rn = np.linalg.norm(r)
    x,y,z = r
    r1 = x+1j*y

    v =  np.array([
        # n = 0
        cs[ 0]      *1./rn ,
        # n = 1
        cs[ 1]*r0   *z   /rn**3   * np.sqrt(3.),
        cs[ 2]*r0   *  r1/rn**3   * np.sqrt(3.),
        # n = 2
        cs[ 3]*r0**2*(3.*z**2       - rn**2) / (2.*rn**5)   * np.sqrt(5.),
        cs[ 4]*r0**2* 3.*z   *r1             /     rn**5    * np.sqrt(5./3.),
        cs[ 5]*r0**2* 3.     *r1**2          /     rn**5    * np.sqrt(5./12.),
        # n = 3
        cs[ 6]*r0**3*(15.*z**2       -  9.*rn**2) * z / (6.*rn**7)   * np.sqrt(7.),
        cs[ 7]*r0**3*(15.*z**2       -  3.*rn**2) *r1 / (2.*rn**7)   * np.sqrt(7./6.),
        cs[ 8]*r0**3* 15.*z   *r1**2                  /     rn**7    * np.sqrt(7./60.),
        cs[ 9]*r0**3* 15.     *r1**3                  /     rn**7    * np.sqrt(7./360.),
        # n = 4
        cs[10]*r0**4*(105.*z**4 - 90.*rn**2*z**2 + 9.*rn**4) / (24.*rn**9)   * np.sqrt(9.   ),
        cs[11]*r0**4*(105.*z**2 - 45.*rn**2) * z * r1 / (6.*rn**9)           * np.sqrt(9./10.),
        cs[12]*r0**4*(105.*z**2 - 15.*rn**2) *  r1**2 / (2.*rn**9)           * np.sqrt(9./180.),
        cs[13]*r0**4* 105.*z                 *  r1**3 /     rn**9            * np.sqrt(9./2520.),
        cs[14]*r0**4* 105.                   *  r1**4 /     rn**9            * np.sqrt(9./20160.)
    ], dtype=np.complex128)
    return mu*v.real.sum()

def geoid_b(a, f):
    b = a*(1-f)
    def func(lmb, phi, h):
        return np.array([
            a**2 * np.cos(lmb) * np.cos(phi),
            a**2 * np.sin(lmb) * np.cos(phi),
            b**2 * np.sin(phi)
        ], dtype=np.float64) / np.sqrt((a*np.cos(phi))**2 + (b*np.sin(phi))**2)
    return func

class GravityVTest(unittest.TestCase):
    def test_v(self):
        r0 = 7000000.
        cs = np.ones(15)
        g = Gravity(4, cs, 1., 1)

        size = 1000
        actual = np.ndarray((size,), dtype=np.float64)
        expected = np.ndarray((size,), dtype=np.float64)

        for i in range(size):
            r = np.array([
                r0*(1 + 10*np.random.rand()),
                r0*(1 + 10*np.random.rand()),
                r0*(1 + 10*np.random.rand())
            ], dtype=np.float64)
            actual[i] = g.potential(r)
            expected[i] = potential(cs, 1, 1, r)

        delta = np.abs(2 * (expected - actual)/(expected + actual))
        flags = delta < 1e-14

        if not flags.all():
            self.fail(msg="\n".join("%2d\t%+.10e\t%+.10e\t%+.10e" % 
                    (i, actual[i], expected[i], delta[i])
                    for i in np.arange(len(delta))[~flags]
                )
            )

    def test_gem4_potential(self):
        gdf = read_gdf(os.path.join(sys.__resources__, "gem4", "gem4-9280.gdf"))
        gfc = read_gfc(os.path.join(sys.__resources__, "gem4", "gem4.gfc"))

        g = Gravity(
            gfc["max_degree"], 
            gfc["data"], 
            gfc["earth_gravity_constant"], 
            gfc["radius"]
        )

        el = geoid_b(gdf["radiusrefpot"], gdf["flatrefpot"])

        DEG = np.pi/180.
        actual = np.ndarray(gdf["data"].shape[0], dtype=np.float64)
        h = gdf["height_over_ell"]
        i = 0
        for p in gdf["data"]:
            actual[i] = g.potential(el(p[0]*DEG, p[1]*DEG, h))
            i += 1

        delta = np.abs(2*(actual-gdf["data"][:,2])/(actual+gdf["data"][:,2]))
        flags = delta < 2e-14 #todo  2
        if not flags.all():
            self.fail(msg="\n".join("%2d\t%+.10e\t%+.10e\t%+.10e" % 
                    (i, actual[i], expected[i], delta[i])
                    for i in np.arange(len(delta))[~flags]
                )
            )

    def _test_egm96_potential(self):
        self.fail("todo. It is very very slow. I am going to use OpenCL to speed up my program")

        gdf = read_gdf(os.path.join(sys.__resources__, "egm96", "egm96-9287.gdf"))
        gfc = read_gfc(os.path.join(sys.__resources__, "egm96", "egm96.gfc"))

File: test/test_gravity.c
#include "gravity.h"

#include <stdio.h>

static int potential44 (gravity_t * gravity, double x, double y, double z, double epsilon) {
    complex_t cs[15] = {};
    gravity->cs = cs;
    const complex_t ones = {1, 1};
    double rn = norm(x, y, z);
    double rd = gravity->r0/rn;
    double rr0[] = {1, rd, rd*rd, rd*rd*rd, rd*rd*rd*rd};
    double mu_d = gravity->mu / rn;
    double zd = z / rn;
    double zz[] = {1, zd, zd*zd, zd*zd*zd, zd*zd*zd*zd};
    complex_t r1d = {x/rn, y/rn};
    complex_t rr1[] = {{1,0}, r1d, r1d, r1d, r1d};
    complex_mult(rr1[2], rr1[1])
    complex_mult(rr1[3], rr1[2])
    complex_mult(rr1[4], rr1[3])

    double expected[] = {
        // m = 0
        mu_d * complex_dot_real(ones, rr1[0]) * rr0[0]*      zz[0],
        mu_d * complex_dot_real(ones, rr1[0]) * rr0[1]*      zz[1]                   * sqrt(3.),
        mu_d * complex_dot_real(ones, rr1[0]) * rr0[2]*(  3.*zz[2]             - 1.) * sqrt(5./4.),
        mu_d * complex_dot_real(ones, rr1[0]) * rr0[3]*( 15.*zz[3] - 9.* zz[1]     ) * sqrt(7.)/6.,
        mu_d * complex_dot_real(ones, rr1[0]) * rr0[4]*(105.*zz[4] - 90.*zz[2] + 9.) * sqrt(9.)/24.,
        // m = 1
        mu_d * complex_dot_real(ones, rr1[1]) * rr0[1]*      zz[0]              * sqrt(3.),
        mu_d * complex_dot_real(ones, rr1[1]) * rr0[2]*   3.*zz[1]              * sqrt(5./3.),
        mu_d * complex_dot_real(ones, rr1[1]) * rr0[3]*( 15.*zz[2] - 3.       ) * sqrt(7./24.),
        mu_d * complex_dot_real(ones, rr1[1]) * rr0[4]*(105.*zz[3] - 45.*zz[1]) * sqrt(9./10.)/6.,
        // m = 2
        mu_d * complex_dot_real(ones, rr1[2]) * rr0[2]*   3.*zz[0]        * sqrt(5./12.),
        mu_d * complex_dot_real(ones, rr1[2]) * rr0[3]*  15.*zz[1]        * sqrt(7./60.),
        mu_d * complex_dot_real(ones, rr1[2]) * rr0[4]*(105.*zz[2] - 15.) * sqrt(9./180.)/2.,
        // m = 3
        mu_d * complex_dot_real(ones, rr1[3]) * rr0[3]*  15.*zz[0] * sqrt(7./360.),
        mu_d * complex_dot_real(ones, rr1[3]) * rr0[4]* 105.*zz[1] * sqrt(9./2520.),
        // m = 4
        mu_d * complex_dot_real(ones, rr1[4]) * rr0[4]* 105.*zz[0] * sqrt(9./20160.)
    };
    double actual = 0;
    double delta = 0;
    int i;
    for (i=0; i<15; i++) {
        complex_set(cs[i], ones)
        actual = potential(gravity, x, y, z);
        delta  = 2 * fabs((actual - expected[i]) / (actual + expected[i]));
        if (delta > epsilon) {
            fprintf(stderr, "#%d delta = %+.12e actual = %+.12e expected = %+.12e\n", i, delta, actual, expected[i]);
            return 1;
        }
        complex_set_zero(cs[i])
    }
    return 0;
}

int test_potential() {
    int res = 0;
    double epsilon = 1e-12;

    const gravity_t * gem4 = gravity_get("GEM4");
    double k[15];
    gravity_t gravity = {4, 1., 1., NULL, k};
    int n,m, i=0, gem4_i;
    for (m=0; m<=4; m++) {
        gem4_i = (2*gem4->n + 3 - m)*m/2;
        for (n=m; n <= 4; n++) {
            fprintf(stderr, "i = %d gem4_i = %d\n", i, gem4_i);
            k[i++] = gem4->k[gem4_i++];
        }
    }
    for (i=0; i<15; i++) fprintf(stderr, "%d %f\n", i, k[i]);

    int size = 2 + 11 * 24;
    struct{double lmb,phi;} angles[size];
    double lmb, phi;
    angles[0].lmb = 0;
    angles[0].phi = - M_PI / 2;
    angles[1].lmb = 0;
    angles[1].phi = M_PI / 2;
    i = 2;
    for (lmb = 0; lmb < 2*M_PI - 0.1; lmb += M_PI/12) {
        for (phi = - 5*M_PI/12; phi < M_PI/2 - 0.1; phi += M_PI/12) {
            angles[i].lmb = lmb;
            angles[i].phi = phi;
            i++;
        }
    }
    double x,y,z;
    for (i=0; i<size; i++) {
        fprintf(stderr, " angles : %f, %f \n", angles[i].lmb, angles[i].phi);
        x = cos(angles[i].phi) * cos(angles[i].lmb);
        y = cos(angles[i].phi) * sin(angles[i].lmb);
        z = sin(angles[i].phi);
        fprintf(stderr, " r = %f, %f, %f \n", x, y, z);
        if (res = potential44(&gravity, x,y,z, epsilon)) return res;
    }
}
